### PR TITLE
Cast constant value in basic_math

### LINKED ISSUE #139: Curse of float64
Adding `int` to 0-dim array causes a `float64` value.

```
>>> (numpy.array(1, numpy.float32) + 1).dtype
dtype('float64')
```

When its dimension is greater than 0, this problem isn't happen.
```
>>> (numpy.array([1], numpy.float32) + 1).dtype
dtype('float32’)
```

Only functions in `basic_math` accept 0-dim array as an input value.
These functions only forward input data to numpy, and it may causes `float64` problem.

We need to check types of arguments and when they contains `float32` and standard python types like `int` or `float`, we need to convert results to `float32` as results of calculation are `float64`.

For example, this code is dangerous:
```
loss = softmax_cross_entropy(...) / 10
loss.backward()
```

### PR DESCRIPTION
**Pull request by unnonouno**
Check and cast all constants that is calculated with `Variable`.

fix #139

### CODE REVIEW

**Comment by unnonouno** [comment]
```
accum_loss = 0  # float
accum_loss += loss.data  # float + float32 = float64
x * accum_loss ...  # float32 * float64 = float64
```
causes float64 problem..

**Comment by unnonouno** [comment]
As `isinstance(numpy.float64(1.0), float) == True`, multiplication to `float64` is safe. I added tests.

**Comment by unnonouno** [comment]
..., but `isinstance(numpy.int64(1), int) == False` only in py3. I use `numpy.isscalar` instead.

**Comment by okuta** [comment]
LGTM!
